lib/graphql/analysis/ast/query_complexity.rb

```diff
@@ -8,113 +8,211 @@ class QueryComplexity < Analyzer
+              # This can happen when several branches are being analyzed together
+              true
+            else
+              # Check if these two scopes have _any_ types in common.
+              possible_right_types = query.possible_types(right_scope)
+              possible_left_types = query.possible_types(left_scope)
+              possible_right_types.each do |rt|
+                if possible_left_types.include?(rt)
+                  return true
+                end
+              end
+              false
             end
-            max
           end
 
-          # Store the complexity for the branch on `type_defn`.
-          # Later we will see if this is the max complexity among branches.
-          def merge(type_defn, key, complexity)
-            @types[type_defn][key] = complexity
-          end
-        end
+          def merged_max_complexity_for_scopes(query, scoped_children_hashes)
+            # Figure out what scopes are possible here.
+            all_scopes = []
+            scoped_children_hashes.each do |h|
+              all_scopes.concat(h.keys)
+            end
+            all_scopes.uniq!
+
+            # If an abstract scope is present, but _all_ of its concrete types
+            # are also in the list, remove it from the list of scopes to check,
+            # because every possible type is covered by a concrete type.
+            # (That is, there are no remainder types to check.)
+            all_scopes.reject! do |scope|
+              scope.kind.abstract? && (
+                query.possible_types(scope).all? { |t| all_scopes.include?(t) }
+              )
+            end
 
-        class ConcreteTypeComplexity
-          attr_reader :max_possible_complexity
+            # This will hold `{ type => int }` pairs, one for each possible branch
+            complexity_by_scope = {}
+
+            # For each scope,
+            # find the lexical selections that might apply to it,
+            # and gather them together into an array.
+            # Then, treat the set of selection hashes
+            # as a set and calculate the complexity for them as a unit
+            all_scopes.each do |scope|
+              # These will be the selections on `scope`
+              children_for_scope = []
+              scoped_children_hashes.each do |sc_h|
+                sc_h.each do |inner_scope, children_hash|
+                  if applies_to?(query, scope, inner_scope)
+                    children_for_scope << children_hash
+                  end
+                end
+              end
+
+              # Calculate the complexity for `scope`, merging all
+              # possible lexical branches.
+              complexity_value = merged_max_complexity(query, children_for_scope)
+              complexity_by_scope[scope] = complexity_value
+            end
 
-          def initialize
-            @max_possible_complexity = 0
+            # Return the max complexity among all scopes
```

Comment:
Nice! So this means that when any number of concrete types might be resolved, we err on the side of picking the most expensive one, complexity-wise?

Reply:
Yes, that's right.